docs/exposing_new_functions.mld

```diff
@@ -0,0 +1,31 @@
+{0 Exposing New Functions to stanc3}
+
+
+
+{1 Background}
+
+For a function to be built into Stan, it has to be included in the Stan Math library and its signature has to be
+exposed to the parser. To do the latter, we have to add a corresponding line in [src/middle/Stan_math_signatures.ml].
+The parser uses the signatures defined there to do the type checking. 
+
+
+{1 Adding a distribution function }
+To add a distribution, we have to find the line containing [let distributions =]. The existing distributions can be
+used for reference. The first argument defines the type of function we want to add.
+The second argument is the function name without the [_...] part. The third argument specifies
+the argument types of the function. The last argument describes whether the return type can be represented as an
+Array of Structs (AoS) or as a Struct of Arrays (SoA). The following line exposes a function [foo_lpdf] which takes in
+two real numbers:
+{[
+  ; ([Lpdf], "foo", [DVReal; DVReal], SoA)
+]}
+
+For overloaded functions, we need to add a signature for each combination of variable types. If [foo_lpdf] admits either
+an integer or a real as its first argument, the proper way to expose the function would be:
+{[
+  ; ([Lpdf], "foo", [DVReal; DVReal], SoA)
+  ; ([Lpdf], "foo", [DVInt; DVReal], SoA)
+]}
```

Comment:
This is the correct way to do this but I don't believe we have any distributions overloaded in this way. We do use it in the normal functions section, like:

```ocaml
 add_unqualified ("multiply", ReturnType UComplex, [UComplex; UComplex], AoS) ;
  add_unqualified ("multiply", ReturnType UInt, [UInt; UInt], SoA) ;
  add_unqualified ("multiply", ReturnType UReal, [UReal; UReal], SoA) ;
  add_unqualified ("multiply", ReturnType UVector, [UVector; UReal], SoA) ;
...
```

Reply:
I'm not sure whether I understand correctly, could you elaborate on this? I wrote this in analogy to the way the `wiener` function is [declared](https://github.com/stan-dev/stanc3/blob/c0a37eff2e19b10163e04c4267dff5031a399761/src/middle/Stan_math_signatures.ml#L305), so these two lines would be inside `distributions`.

Reply:
My point was that this list of distributions uses the types like `DVReal` to automatically generate overrides, there are not any distributions which explictly define an override as you do in this example.

So, for something like the `wiener` distribution, all 1024 available signatures are generated only by one line. This is different from the 'normal' functions like `multiply(x,y)`, which need to have each signature written out for each overload

Reply:
Ok, I understand, thanks for the clarification.